NanoAnalysis/python/RecoProbFiller.py

```diff
@@ -0,0 +1,225 @@
+from PhysicsTools.NanoAODTools.postprocessing.framework.datamodel import Collection
+from  ZZAnalysis.NanoAnalysis.initializeMELA import check_enum
+from ZZAnalysis.NanoAnalysis.ZZExtraFiller import *
+import Mela
+
+
+class RecoProbFiller(Module):
+    """Calculates proabilities with Reco-level information. 
+    MELA = Pointer to MELA passed from nanoZZ4lAnalysis.py 
+    """
+    
+    def __init__(self, MELA, NANOVERSION, settingsDict = None):
+        print("***RecoProbFiller", flush=True)
+        self.MELA = MELA
+        self.MELAsettings = settingsDict
+        self.NANOVERSION = NANOVERSION
+            
+    def beginFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
+        self.out = wrappedOutputTree
+        if self.MELAsettings != None: 
+            for p, prob in enumerate(self.MELAsettings): 
+                self.out.branch(f"RecoMela_{prob['Name']}", "F", lenVar = "nZZCand", limitedPrecision=16, title="User-defined Reco-level probability")
+                if prob.get("ispm4l", False): 
+                    self.out.branch("RecoMela_"+prob["Name"]+"_ScaleUp", "F", lenVar = "nZZCand", limitedPrecision=16, title="User-defined Reco-level m4l probability with Scale uncertainties up")
+                    self.out.branch("RecoMela_"+prob["Name"]+"_ScaleDown", "F", lenVar = "nZZCand", limitedPrecision=16, title="User-defined Reco-level m4l probability with Scale uncertainties down")
+                    self.out.branch("RecoMela_"+prob["Name"]+"_SystUp", "F", lenVar = "nZZCand", limitedPrecision=16, title="User-defined Reco-level m4l probability with Systematic uncertainties up")
+                    self.out.branch("RecoMela_"+prob["Name"]+"_SystDown", "F", lenVar = "nZZCand", limitedPrecision=16, title="User-defined Reco-level m4l probability with Systematic uncertainties down")
+                if prob["computeprop"]: 
+                    self.out.branch("RecoMela_"+prob["Name"]+"_prop", "F", lenVar = "nZZCand", limitedPrecision=16, title="User-defined Reco-level probability with non-default propagator scheme")
+                        
+
+        
+    def analyze(self, event):
+        # Only run if we have a pyfragment 
+        if self.MELAsettings != None: 
+            cands = Collection(event, 'ZZCand')
+            electrons = Collection(event, "Electron")
+            muons = Collection(event, "Muon")
+            self.leps = list(electrons) + list(muons)
+            fsrPhotons = Collection(event, "FsrPhoton")
+            
+            mothers = Mela.SimpleParticleCollection_t()
+            daughters = Mela.SimpleParticleCollection_t()
+            associated = Mela.SimpleParticleCollection_t()
+
+            
+
+            ### Sort the MELASettings dictionary so that all probabilities with divideP are last 
+            sortedSettings = []
+            denominator_name = ""
+            for p, prob in enumerate(self.MELAsettings): 
```

Comment:
same comment as for LHEAngProbFiller  - sorting can be done once in the constructor